Add versions to banner

matrei opened the pull request:
Will by default output versions below banner like:
```console
    >                                                                            <
    >                            ____           _ _                              <
    >                           / ___|_ __ __ _(_) |___                          <
    >                          | |  _| '__/ _` | | / __|                         <
    >                          | |_| | | | (_| | | \__ \                         <
    >                           \____|_|  \__,_|_|_|___/                         <
    >                          https://grails.apache.org/                        <
    >                                                                            <
            grails-test-examples-async-events-pubsub-demo: 7.1.0-SNAPSHOT         
           JVM: BellSoft 17.0.17 | Grails: 7.1.0-SNAPSHOT | Groovy: 4.0.29        
                         Spring Boot: 3.5.7 | Spring: 6.2.12
```

See documentation changes in this PR for configuration options.

matrei committed:
feat: add versions to banner

sbglasius reviewed (approved):
What a great idea 😄

sbglasius commented:
Can the java-version show what distribution it comes from (Amazon, Belsoft, Oracle, etc?)

To answer myself: `System.getProperty('java.vendor')`

sbglasius commented:
Something like this: https://github.com/apache/grails-core/pull/15230

sbglasius committed:
Include Java vendor in Grails banner version info (#15230)

matrei committed:
fix: rework `printBanner`

No need to store the versions or use resolvers.
They are only used in the `printBanner` method.

codeconsole commented:
looks nice, but I wonder I wonder if it should also display when banner is off?  In all my apps, I have been using
```groovy
log.info "${Metadata.current['info.app.name']}  v${Metadata.current['info.app.version']} (Java ${System.getProperty('java.version')} Groovy ${GroovySystem.getVersion()} Grails ${Metadata.current['info.app.grailsVersion']} Spring Boot ${SpringBootVersion.version} Spring ${SpringVersion.version})"
```
because it is useful for troubleshooting to know what versions you are running.   I have banner off because I don't want to fill my logs with a banner (especially when I am running multiple instances).    

Or have a separate setting to display it if banner is off?

sbglasius commented on grails-core/src/main/groovy/grails/boot/GrailsBanner.groovy:
```diff
@@ -0,0 +1,92 @@
+                'Groovy': GroovySystem.version,
+                'Spring Boot': SpringBootVersion.version,
+                'Spring': SpringVersion.version
+        ]
```
Suggestion: If this was split out into a `protected` method, it would be possible for anyone to extend the `GrailsBanner` and the list, with information from CI/CD etc.

matrei replied:
@sbglasius See updated PR description.

matrei committed:
feat: make GrailsBanner configurable

- Allow overriding behavior
- Add conditional output of art and versions

matrei commented:
> looks nice, but I wonder I wonder if it should also display when banner is off? In all my apps, I have been using
> 
> ```groovy
> log.info "${Metadata.current['info.app.name']}  v${Metadata.current['info.app.version']} (Java ${System.getProperty('java.version')} Groovy ${GroovySystem.getVersion()} Grails ${Metadata.current['info.app.grailsVersion']} Spring Boot ${SpringBootVersion.version} Spring ${SpringVersion.version})"
> ```
> 
> because it is useful for troubleshooting to know what versions you are running. I have banner off because I don't want to fill my logs with a banner (especially when I am running multiple instances).
> 
> Or have a separate setting to display it if banner is off?

@codeconsole See updated PR description, will this accommodate your use case?

matrei committed:
feat: allow overriding banner versions formatting

codeconsole commented:
> @codeconsole See updated PR description, will this accommodate your use case?

yes 😄

codeconsole commented:
@matrei bonus points if we can get Spring Security in there? 
```groovy
Class.forName('org.springframework.security.core.SpringSecurityCoreVersion')?.version
```

sbglasius reviewed (approved):
Awesome!

matrei commented:
> bonus points if we can get Spring Security in there?
> 
> ```groovy
> Class.forName('org.springframework.security.core.SpringSecurityCoreVersion')?.version
> ```

@codeconsole I'm not sure we should do class loading for optional dependencies in the banner. You could add it yourself in your app:
```groovy
    static void main(String[] args) {
        new GrailsApp(Application).tap {
            banner = new GrailsBanner() {
                @Override
                Map<String, String> createBannerVersions(Environment environment) {
                    super.createBannerVersions(environment) +
                        ['Spring Security': Class.forName('org.springframework.security.core.SpringSecurityCoreVersion')?.version]
                }
            }
        }.run()
    }
```

jdaugherty reviewed (approved):
I love the idea of this, but can we add documentation for this behavior?  I think most of that documentation is already in this PR description - it just needs added to the docs

matrei committed:
feat: add map constructor to `GrailsBanner`

codeconsole commented:
> @codeconsole I'm not sure we should do class loading for optional dependencies in the banner. You could add it yourself in your app

I agree for any other optional dependencies, but Spring Security is such a fundamental core dependency that it would be nice to display as a default.  You can just surround it with an ignored ClassNotFoundException and it is harmless.

sbglasius commented:
@matrei Since SpringSecurityCore is under Apache Grails, I agree with @codeconsole, but in any other case I would disagree too.

Now that this is an open issue, how about creating a `META-INF/service/grails.banner.loader` with the name of the `GrailsBanner` class implemented, and then (I assume) SpringSecurityCore could extend `GrailsBanner` and override `META-INFO/service/grails.banner.loader` I'm not 100% sure about the load order in such case? and perhaps it's just a big overkill 😀

matrei committed:
feat: add more banner customization options

matrei committed:
docs: document the banner functionality

matrei committed:
style: remove unnecessary semicolons

matrei commented:
> I love the idea of this, but can we add documentation for this behavior? I think most of that documentation is already in this PR description - it just needs added to the docs

@jdaugherty The functionality is now documented, thanks for reminding me!

matrei commented:
@codeconsole and @sbglasius 
I have added support for optional dependency versions via configuration:
```yaml
grails:
  banner:
    versions:
      include:
        - spring-security
        - tomcat
        - jetty
        - undertow
```

sbglasius reviewed (approved):
Marvelous
